### PR TITLE
[Frontend] Support OpenAI batch file format

### LINKED ISSUE #4777: [Feature]: Support the OpenAI Batch Chat Completions file format
### 🚀 The feature, motivation and pitch

I'm working on a use case that involves running a the same dataset/prompt across multiple models (including some OpenAI and some open source models). I would like to be able to do batch inference on many requests in a file that follow the [OpenAI Batch file format](https://platform.openai.com/docs/guides/batch/1-preparing-your-batch-file).
1. This follows the spirit/pattern of the popular openai api server interface for using vllm.
2. It is easy to adapt existing code that calls web endpoints to generate these files (since the `body` field is essentially what you would pass into web endpoint).
3. This format doesn't require the user to think about rate limits, parallelism, etc.

I'll lay out an implementation plan here, which **I'm willing to contribute an implementation for**.

# Interface

The primary interface would be via CLI command.

```
$ python -m vllm.entrypoints.openai_batch --help
Usage: openai_batch [OPTIONS]

  Run offline inference on a file which conforms to the OpenAI Batch file format. https://platform.openai.com/docs/guides/batch/getting-started

Options:
  --version  Show the version and exit.
  --help     Show this message and exit.
  -i
  --input-file The path or url to a single input file. Currently supports local file paths, or the http protocol (http or https). If a URL is specified, the file should be available via HTTP GET.
  -o
  --output-file The path or url to a single output file. Currently supports local file paths, or web (http or https) urls. If a URL is specified, the file should be available via HTTP PUT.

Exit status:
  0 No problems occured.
  1 Generic error code.
```

# Implementation

The feature of this feature should be a fairly independent wrapper around the vLLM core, and the implementation shouldn't be very involved. I propose only a very minor cleanup to `OpenAIServingChat` in which `OpenAIServingChat::create_chat_completion` function signature will change to (the only change is the `raw_request` is removed):

```
    async def create_chat_completion(
        self, request: ChatCompletionRequest, is_aborted : Optional[Callable[[], Awaitable[bool]] = None
    ) -> Union[ErrorResponse, AsyncGenerator[str, None], ChatCompletionResponse]:
```

The only caller of the function (`api_server.py`) will simply pass `is_aborted = raw_request.is_disconnected`.

The rest of the implementation will be:
1. Create a new pydantic model for the request.
2. Load the local file or url into a list of request objects.
3. Submit the requests to the openai_serving_chat.
4. Write the outputs.

### Alternatives

It's possible to simply instantiate the api server and manage parallelizing requests today, but it adds complication to user code since request parallelization is important and difficult to get right and ensuring the http server's port is available isn't always trivial in failure conditions.

Alternative apis:
* Python API: Nothing in this proposal prevents or makes it more difficult to introduce a python api later, though there are additional interface decisions to make (async vs synchronous, exclusive vs shared engine, etc).
* REST API: The openai REST API using an `input_file_id` field. It's not obvious to me how implementing this field in the same way as openai (with validation, etc) should interact with the rest of the vLLM project, so I will leave it out of scope for now (perhaps others can chip in if this is a desired feature). 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wuisawesome**
This PR adds basic support for the OpenAI batch file format (for chat endpoints).

The following help page summarizes the usage options:

```
# python -m vllm.entrypoints.openai.run_batch --help
usage: run_batch.py [-h] -i INPUT_FILE -o OUTPUT_FILE [--response-role RESPONSE_ROLE] [--model MODEL] [--tokenizer TOKENIZER] [--skip-tokenizer-init] [--revision REVISION]
                    [--code-revision CODE_REVISION] [--tokenizer-revision TOKENIZER_REVISION] [--tokenizer-mode {auto,slow}] [--trust-remote-code] [--download-dir DOWNLOAD_DIR]
                    [--load-format {auto,pt,safetensors,npcache,dummy,tensorizer}] [--dtype {auto,half,float16,bfloat16,float,float32}] [--kv-cache-dtype {auto,fp8}]
                    [--quantization-param-path QUANTIZATION_PARAM_PATH] [--max-model-len MAX_MODEL_LEN] [--guided-decoding-backend {outlines,lm-format-enforcer}] [--worker-use-ray]
                    [--pipeline-parallel-size PIPELINE_PARALLEL_SIZE] [--tensor-parallel-size TENSOR_PARALLEL_SIZE] [--max-parallel-loading-workers MAX_PARALLEL_LOADING_WORKERS]
                    [--ray-workers-use-nsight] [--block-size {8,16,32}] [--enable-prefix-caching] [--use-v2-block-manager] [--num-lookahead-slots NUM_LOOKAHEAD_SLOTS] [--seed SEED]
                    [--swap-space SWAP_SPACE] [--gpu-memory-utilization GPU_MEMORY_UTILIZATION] [--num-gpu-blocks-override NUM_GPU_BLOCKS_OVERRIDE]
                    [--max-num-batched-tokens MAX_NUM_BATCHED_TOKENS] [--max-num-seqs MAX_NUM_SEQS] [--max-logprobs MAX_LOGPROBS] [--disable-log-stats]
                    [--quantization {aqlm,awq,fp8,gptq,squeezellm,gptq_marlin,marlin,deepspeedfp,None}] [--enforce-eager] [--max-context-len-to-capture MAX_CONTEXT_LEN_TO_CAPTURE]
                    [--max-seq_len-to-capture MAX_SEQ_LEN_TO_CAPTURE] [--disable-custom-all-reduce] [--tokenizer-pool-size TOKENIZER_POOL_SIZE]
                    [--tokenizer-pool-type TOKENIZER_POOL_TYPE] [--tokenizer-pool-extra-config TOKENIZER_POOL_EXTRA_CONFIG] [--enable-lora] [--max-loras MAX_LORAS]
                    [--max-lora-rank MAX_LORA_RANK] [--lora-extra-vocab-size LORA_EXTRA_VOCAB_SIZE] [--lora-dtype {auto,float16,bfloat16,float32}] [--max-cpu-loras MAX_CPU_LORAS]
                    [--fully-sharded-loras] [--device {auto,cuda,neuron,cpu}] [--image-input-type {pixel_values,image_features}] [--image-token-id IMAGE_TOKEN_ID]
                    [--image-input-shape IMAGE_INPUT_SHAPE] [--image-feature-size IMAGE_FEATURE_SIZE] [--scheduler-delay-factor SCHEDULER_DELAY_FACTOR] [--enable-chunked-prefill]
                    [--speculative-model SPECULATIVE_MODEL] [--num-speculative-tokens NUM_SPECULATIVE_TOKENS] [--speculative-max-model-len SPECULATIVE_MAX_MODEL_LEN]
                    [--speculative-disable-by-batch-size SPECULATIVE_DISABLE_BY_BATCH_SIZE] [--ngram-prompt-lookup-max NGRAM_PROMPT_LOOKUP_MAX]
                    [--ngram-prompt-lookup-min NGRAM_PROMPT_LOOKUP_MIN] [--model-loader-extra-config MODEL_LOADER_EXTRA_CONFIG]
                    [--served-model-name SERVED_MODEL_NAME [SERVED_MODEL_NAME ...]] [--engine-use-ray] [--disable-log-requests] [--max-log-len MAX_LOG_LEN]

vLLM OpenAI-Compatible batch runner.

options:
  -h, --help            show this help message and exit
  -i INPUT_FILE, --input-file INPUT_FILE
                        The path or url to a single input file. Currently supports local file paths, or the http protocol (http or https). If a URL is specified, the file should be
                        available via HTTP GET.
  -o OUTPUT_FILE, --output-file OUTPUT_FILE
                        The path or url to a single output file. Currently supports local file paths, or web (http or https) urls. If a URL is specified, the file should be
                        available via HTTP PUT.
  --response-role RESPONSE_ROLE
                        The role name to return if `request.add_generation_prompt=true`.
  --model MODEL         Name or path of the huggingface model to use.
...
```

Assuming the following file exists:

```
# cat ../batch.jsonl 
{"custom_id": "request-1", "method": "POST", "url": "/v1/chat/completions", "body": {"model": "meta-llama/Meta-Llama-3-8B-Instruct", "messages": [{"role": "system", "content": "You are a helpful assistant."},{"role": "user", "content": "Hello world!"}],"max_tokens": 1000}}
{"custom_id": "request-2", "method": "POST", "url": "/v1/chat/completions", "body": {"model": "meta-llama/Meta-Llama-3-8B-Instruct", "messages": [{"role": "system", "content": "You are an unhelpful assistant."},{"role": "user", "content": "Hello world!"}],"max_tokens": 1000}}
```

The program is run as `python -m vllm.entrypoints.openai.run_batch -i ../batch.jsonl -o ../results.jsonl --model meta-llama/Meta-Llama-3-8B-Instruct` (this produces the typical logging from the async engine, which I've omitted).

Finally, the resulting file produced is:

```
# cat ../results.jsonl
{"id":"vllm-383d1c59835645aeb2e07d004d62a826","custom_id":"request-1","response":{"id":"cmpl-61c020e54b964d5a98fa7527bfcdd378","object":"chat.completion","created":1715633336,"model":"meta-llama/Meta-Llama-3-8B-Instruct","choices":[{"index":0,"message":{"role":"assistant","content":"Hello! It's great to meet you! I'm here to help with any questions or tasks you may have. What's on your mind today?"},"logprobs":null,"finish_reason":"stop","stop_reason":null}],"usage":{"prompt_tokens":25,"total_tokens":56,"completion_tokens":31}},"error":null}
{"id":"vllm-42e3d09b14b04568afa3f1797751a267","custom_id":"request-2","response":{"id":"cmpl-f44d049f6b3a42d4b2d7850bb1e31bcc","object":"chat.completion","created":1715633336,"model":"meta-llama/Meta-Llama-3-8B-Instruct","choices":[{"index":0,"message":{"role":"assistant","content":"*silence*"},"logprobs":null,"finish_reason":"stop","stop_reason":null}],"usage":{"prompt_tokens":27,"total_tokens":32,"completion_tokens":5}},"error":null}
```

FIX #4777

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/entrypoints/openai/api_server.py
Comment (robertgshaw2-neuralmagic) [inline vllm/entrypoints/openai/api_server.py:104]: i would prefer if this PR did not touch this file

File: vllm/entrypoints/openai/protocol.py
Comment (robertgshaw2-neuralmagic) [inline vllm/entrypoints/openai/protocol.py:513]: Is there a reason we cannot support `/v1/embeddings`?

https://platform.openai.com/docs/guides/batch

File: vllm/entrypoints/openai/serving_chat.py
Comment (robertgshaw2-neuralmagic) [inline vllm/entrypoints/openai/serving_chat.py:102]: I do not think we should change the signature of this function in the PR

File: vllm/entrypoints/openai/run_batch.py
Comment (robertgshaw2-neuralmagic) [inline vllm/entrypoints/openai/run_batch.py:121]: I assume this lets each request run concurrently right?

File: tests/entrypoints/test_openai_run_batch.py
Comment (robertgshaw2-neuralmagic) [inline tests/entrypoints/test_openai_run_batch.py:12]: It there any way we can add a test that the input / output formats conform to the openai api

File: vllm/entrypoints/openai/protocol.py
Comment (robertgshaw2-neuralmagic) [reply vllm/entrypoints/openai/protocol.py:513]: [ This could be done in a follow up PR ]

**Review by robertgshaw2-redhat — COMMENTED** [review]
I think this is a useful feature. Thanks for doing it!

My only concern is that this is not really an OpenAI-compatible API - but rather an OpenAI-compatible file format. I think this is fine, but perhaps we should make this explicit in the documentation ...

See comments in code - my biggest feedback is around the `is_aborted` - I would rather this PR not touch that function signature.

We should add embedding models as a follow up

File: vllm/entrypoints/openai/serving_chat.py
Comment (wuisawesome) [reply vllm/entrypoints/openai/serving_chat.py:102]: Sure, are you just trying to keep this method signature backwards compatible?

2 possibilities 
1) Make `raw_request: Optional[Request]` and default to never aborting if `raw_request is None`
2) I can make a fake `RawRequest` in the `run_batch.py` to pass in.

I'm fine either way, or if you had something else in mind. Let me know what you prefer.

File: vllm/entrypoints/openai/protocol.py
Comment (wuisawesome) [reply vllm/entrypoints/openai/protocol.py:513]: Yes, we can support the other 2 endpoints in follow ups.

File: vllm/entrypoints/openai/run_batch.py
Comment (wuisawesome) [reply vllm/entrypoints/openai/run_batch.py:122]: yep, that way the async engine can decide how to queue/batch things.

File: tests/entrypoints/test_openai_run_batch.py
Comment (wuisawesome) [reply tests/entrypoints/test_openai_run_batch.py:15]: sure

**Comment by robertgshaw2-neuralmagic** [comment]
This looks great.

Can you just add a short example of `run_batch.py` in `/examples`

File: examples/offline_inference_openai.md
Comment (wuisawesome) [inline examples/offline_inference_openai.md:9]: These examples links point to master, so they don't work now, but should work once merged.

**Review by robertgshaw2-redhat — APPROVED** [review]

**Comment by robertgshaw2-neuralmagic** [comment]
LGTM. Waiting for the CI to pass

**Comment by robertgshaw2-neuralmagic** [comment]
Thanks for the good work here

**Comment by AllenDou** [comment]
openi_example_batch.jsonl -> openai_example_batch.jsonl ?

**Comment by wuisawesome** [comment]
🤦yes, thanks for catching that.

**Comment by robertgshaw2-neuralmagic** [comment]
@wuisawesome ping me on slack and ill merge it

**Comment by wuisawesome** [comment]
I think we're good here, Sang merged it already

**Pull request merged**
